addons/mrp/models/mrp_workorder.py

```diff
@@ -150,19 +149,31 @@ def _read_group_workcenter_id(self, workcenters, domain):
                                      domain="[('allow_workorder_dependencies', '=', True), ('id', '!=', id), ('production_id', '=', production_id)]",
                                      copy=False)
 
-    @api.depends('production_availability', 'blocked_by_workorder_ids.state', 'qty_ready', 'product_uom_id')
+    @api.depends('qty_ready', 'blocked_by_workorder_ids.state')
     def _compute_state(self):
         for workorder in self:
-            if not workorder.product_uom_id or workorder.state not in ('pending', 'waiting', 'ready'):
+            if not workorder.product_uom_id or workorder.state not in ('blocked', 'ready'):
                 continue
-            blocked = any(w.state not in ('done', 'cancel') for w in workorder.blocked_by_workorder_ids)
             has_qty_ready = float_compare(workorder.qty_ready, 0, precision_rounding=workorder.product_uom_id.rounding) > 0
-            if blocked and not has_qty_ready:
-                workorder.state = 'pending'
-            elif workorder.production_availability == 'assigned' or (has_qty_ready and workorder.blocked_by_workorder_ids):
-                workorder.state = 'ready'
+            if not workorder.blocked_by_workorder_ids or has_qty_ready or all(wo.state in ('cancel', 'done') for wo in workorder.blocked_by_workorder_ids):
+                workorder.write({'state': 'ready'})
             else:
-                workorder.state = 'waiting'
+                workorder.write({'state': 'blocked'})
+
+    def set_state(self, state):
```

Comment:
it could be an inverse function no? (Not sure it's a good idea since it could trigger unwanted side effect)